README.md
# Unique-Unit-Lines
Civilization 5 mod. Applies unique promotions from a unique unit to the whole unit line for the civilization.


Unique Unit Line/Changes.txt
--------------------------------------------------------------
v1
--------------------------------------------------------------
- Created.

--------------------------------------------------------------
v2
--------------------------------------------------------------
- Added missing support for air units.

--------------------------------------------------------------
v3
--------------------------------------------------------------
- Promotions are now only granted if the technology unlocking the unique unit has been researched. Can be disabled by changing UNIQUE_UNIT_LINES_REQUIRE_TECH in UniqueUnitLines_GameDefines.sql.

Unique Unit Line/Lua/UniqueUnitLines_Functions.lua
-- UniqueUnitLines_Functions
-- Author: HungryForFood
-- DateCreated: 3/22/2020 2:58:07 PM
print("UniqueUnitLines_Functions.lua has loaded.")
--=======================================================================================================================
-- INCLUDES
--=======================================================================================================================
include("UniqueUnitLines_Utilities.lua")
--=======================================================================================================================
-- GLOBALS
--=======================================================================================================================
------------------------------------------------------------------------------------------------------------------------
local g_TableInsert = table.insert
-- exceptions for promotions which may be too overpowered early on
local g_tDisallowedPromotionsToGive = {} -- to use, just add to this table, eg: {"Promotion A", "Promotion B", etc}
if g_bRequireTech == true then
	g_TableInsert(g_tDisallowedPromotionsToGive, "PROMOTION_RECON_BANDEIRANTES") -- the Bandeirantes yields are just far too much for early game
end
--=======================================================================================================================
-- GAMES DEFINES
--=======================================================================================================================
------------------------------------------------------------------------------------------------------------------------
local iMaxMajorCivs = GameDefines["MAX_MAJOR_CIVS"]
local iCivilizationBarbarian = GameInfoTypes["CIVILIZATION_BARBARIAN"]
local iCivilizationMinor = GameInfoTypes["CIVILIZATION_MINOR"]
--=======================================================================================================================
-- CACHED TABLES
--=======================================================================================================================
-------------------------------------------------------------------------------------------------------------------------
local g_tPromotionsToAdd = {}
local g_tPromotionsToRemove = {}
--=======================================================================================================================
-- CORE FUNCTIONS
--=======================================================================================================================
-------------------------------------------------------------------------------------------------------------------------
-- LOADING INTO GAME
-------------------------------------------------------------------------------------------------------------------------
function UniqueUnitLines_SequenceGameInitComplete()
	for iPlayer = 0, iMaxMajorCivs - 1 do
		local tPromotionsToAdd = {}
		local tPromotionsToRemove = {}
		local pPlayer = Players[iPlayer]
		if pPlayer:IsEverAlive() then
			local iCivilization = pPlayer:GetCivilizationType()
			
			if iCivilization ~= iCivilizationBarbarian and iCivilization ~= iCivilizationMinor then
				local tUniqueUnitLinesPromotions = {} -- promotions to add to the unit line
				local tDefaultUnitLinePromotions = {} -- promotions to remove from the unit line
				-- find this civ's unique unit and its equivalent default unit
				for tUniqueUnit in DB.Query("SELECT uc.DefaultUnit, uco.UnitType, u.CombatClass, u.IsMounted FROM Civilization_UnitClassOverrides uco, Civilizations c, UnitClasses uc, Units u WHERE c.ID = " .. iCivilization .. " AND uco.CivilizationType = c.Type AND uc.Type = uco.UnitClassType AND u.Type = uc.DefaultUnit;") do 	
					local sUniqueUnit = tUniqueUnit.UnitType
					local sDefaultUnit = tUniqueUnit.DefaultUnit
					local sCombatClass = tUniqueUnit.CombatClass
					local bMounted = false
					if tUniqueUnit.IsMounted > 0 then
						bMounted = true
					end

					local sUnitLine = Unit_GetUnitLine(sCombatClass, bMounted)
					
					if sUnitLine ~= nil then
						-- cache the unique unit promotions
						if tUniqueUnitLinesPromotions[sUnitLine] == nil then
							tUniqueUnitLinesPromotions[sUnitLine] = {}
						end
						
						---- prerequisite tech, not doing this in the first query as we want the unique unit's tech rather than the default unit's tech
						local sPrereqTech = nil
						local iPrereqTech = nil
						if g_bRequireTech == true then
							for tUniqueUnitTech in DB.Query("SELECT PrereqTech FROM Units WHERE Type = \"" .. sUniqueUnit .. "\";") do
								sPrereqTech = tUniqueUnitTech.PrereqTech
							end
							if sPrereqTech ~= nil and sPrereqTech ~= "" then
								iPrereqTech = GameInfoTypes[sPrereqTech]
							end
						end
						
						---- disallowed promotions
						local sDisallowedPromotions = ""
						for i, sPromotion in ipairs(g_tDisallowedPromotionsToGive) do
							if sDisallowedPromotions ~= "" then
								sDisallowedPromotions = sDisallowedPromotions .. ", " .. sPromotion
							else
								sDisallowedPromotions = sDisallowedPromotions .. sPromotion
							end
						end
						
						local sQuery = "SELECT p.ID FROM Unit_FreePromotions fp, UnitPromotions p WHERE fp.UnitType = \"".. sUniqueUnit .."\" AND fp.PromotionType NOT IN(SELECT PromotionType FROM Unit_FreePromotions WHERE UnitType = \"".. sDefaultUnit .."\") AND p.LostWithUpgrade = 0 AND p.Type = fp.PromotionType"
						if sDisallowedPromotions ~= "" then
							sQuery = sQuery .. " AND p.Type NOT IN(\"".. sDisallowedPromotions .."\")"
						end
						sQuery = sQuery .. ";"
						
						for tUniquePromotion in DB.Query(sQuery) do
							g_TableInsert(tUniqueUnitLinesPromotions[sUnitLine], {tUniquePromotion.ID, iPrereqTech})
						end
						
						
						-- cache the default unit promotions
						if tDefaultUnitLinePromotions[sUnitLine] == nil then
							tDefaultUnitLinePromotions[sUnitLine] = {}
						end
						
						sQuery = "SELECT p.ID FROM Unit_FreePromotions fp, UnitPromotions p WHERE UnitType = \"".. sDefaultUnit .."\" AND PromotionType NOT IN(SELECT PromotionType FROM Unit_FreePromotions WHERE UnitType = \"".. sUniqueUnit .."\") AND p.Type = fp.PromotionType;"
						for tDefaultPromotion in DB.Query(sQuery) do
							g_TableInsert(tDefaultUnitLinePromotions[sUnitLine], {tDefaultPromotion.ID, iPrereqTech})
						end
					end
				end
				
				g_tPromotionsToAdd[iPlayer] = tUniqueUnitLinesPromotions
				g_tPromotionsToRemove[iPlayer] = tDefaultUnitLinePromotions
			end
			
			-- grant the promotions to the starting units, only do this for turn 0
			if Game.GetElapsedGameTurns() == 0 then
				for pUnit in pPlayer:Units() do
					local iUnit = pUnit:GetID()
					Unit_DoUniqueUnitLinePromotions(iPlayer, iUnit, g_tPromotionsToAdd[iPlayer], g_tPromotionsToRemove[iPlayer])
				end
			end
		end
	end
end
Events.SequenceGameInitComplete.Add(UniqueUnitLines_SequenceGameInitComplete)
-------------------------------------------------------------------------------------------------------------------------
-- UNIT CREATED
-------------------------------------------------------------------------------------------------------------------------
function UniqueUnitLines_UnitCreated(iPlayer, iUnit, iUnitType, iPlotX, iPlotY)
	Unit_DoUniqueUnitLinePromotions(iPlayer, iUnit, g_tPromotionsToAdd[iPlayer], g_tPromotionsToRemove[iPlayer])
end
GameEvents.UnitCreated.Add(UniqueUnitLines_UnitCreated) -- note that the UnitCreated hook does not work with unit upgrading as the CvUnit::convert() function is called after unit init, replacing all the promotion changes we have done
-------------------------------------------------------------------------------------------------------------------------
-- UNIT UPGRADED
-------------------------------------------------------------------------------------------------------------------------
function UniqueUnitLines_UnitConverted(iOldPlayer, iNewPlayer, iOldUnit, iNewUnit, bIsUpgrade)
	Unit_DoUniqueUnitLinePromotions(iNewPlayer, iNewUnit, g_tPromotionsToAdd[iNewPlayer], g_tPromotionsToRemove[iNewPlayer])
end
GameEvents.UnitConverted.Add(UniqueUnitLines_UnitConverted) -- this covers everything that the UnitCreated hook doesn't, eg upgrades and barbarian captures
-------------------------------------------------------------------------------------------------------------------------
-- TECHNOLOGY RESEARCHED
-------------------------------------------------------------------------------------------------------------------------
function UniqueUnitLines_TeamTechResearched(iTeam, iTech, iChange)
	if iChange > 0 then
		for iPlayer = 0, iMaxMajorCivs - 1 do
			local pPlayer = Players[iPlayer]
			if pPlayer:IsEverAlive() and pPlayer:GetTeam() == iTeam then
				for pUnit in pPlayer:Units() do
					local iUnit = pUnit:GetID()
					Unit_DoUniqueUnitLinePromotions(iPlayer, iUnit, g_tPromotionsToAdd[iPlayer], g_tPromotionsToRemove[iPlayer])
				end
			end
		end
	end
end
GameEvents.TeamTechResearched.Add(UniqueUnitLines_TeamTechResearched) -- when researchiug the unique unit's technology
--==========================================================================================================================
--==========================================================================================================================

Unique Unit Line/Lua/Utilities/UniqueUnitLines_Utilities.lua
-- UniqueUnitLine_Utilities
-- Author: HungryForFood
-- DateCreated: 3/22/2020 2:58:32 PM
--=======================================================================================================================
if (not GameInfo) then return end
--=======================================================================================================================
-- INCLUDES
--=======================================================================================================================
--=======================================================================================================================
-- GLOBALS
--=======================================================================================================================
------------------------------------------------------------------------------------------------------------------------
-- do we consider if we have unlocked the unique unit before granting the its promotions?
g_bRequireTech = false
for tRequireTech in DB.Query("SELECT Value FROM COMMUNITY WHERE Type = \"UNIQUE_UNIT_LINES_REQUIRE_TECH\"") do
	if tRequireTech.Value == 1 then
		g_bRequireTech = true
	end
end
--=======================================================================================================================
-- CACHED TABLES
--=======================================================================================================================
-------------------------------------------------------------------------------------------------------------------------
-- g_tUnitCombats
-- this is just the ID and Type from UnitCombatInfos, caching for speed as this will be checked often
local g_tUnitCombats = {}
for row in DB.Query("SELECT ID, Type FROM UnitCombatInfos;") do 	
	g_tUnitCombats[row.ID] = row.Type
end

-- g_tUnitClasses
-- this is just the ID and Type from UnitClasses, caching for speed as this will be checked often
local g_tUnitClasses = {}
for row in DB.Query("SELECT ID, Type FROM UnitClasses;") do 	
	g_tUnitClasses[row.ID] = row.Type
end
--=======================================================================================================================
-- GAME DEFINES
--=======================================================================================================================
------------------------------------------------------------------------------------------------------------------------
local iCivilizationBarbarian = GameInfoTypes["CIVILIZATION_BARBARIAN"]
local iCivilizationMinor = GameInfoTypes["CIVILIZATION_MINOR"]
local iCivilizationByzantium = GameInfoTypes["CIVILIZATION_BYZANTIUM"]

local iTechChivalry = GameInfoTypes["TECH_CHIVALRY"]

local iPromotionSmallCityPenalty = GameInfoTypes["PROMOTION_SMALL_CITY_PENALTY"]

local iCombatClassMounted = GameInfoTypes["UNITCOMBAT_MOUNTED"]
--=======================================================================================================================
-- LOCAL UTILS
--=======================================================================================================================
-- UNITS
-------------------------------------------------------------------------------------------------------------------------
function Unit_GetUnitLine(sCombatClass, bMounted)
	local sUnitLine = nil
	-- find out which unit line does a combat class belong to
	---- if blank, then it is probably a civilian unit, which usually do not have an upgrade line
	---- though in VP, all the non-great people civilian types have a unit combat assigned already anyway
	if sCombatClass == nil or sCombatClass == "" then
		return sUnitLine
	---- check for the few most common scenarios first
	elseif sCombatClass == "UNITCOMBAT_ARCHER" then
		if bMounted == true then
			sUnitLine = "Mounted Ranged"
		else
			sUnitLine = "Ranged"
		end
	elseif sCombatClass == "UNITCOMBAT_MELEE" or sCombatClass == "UNITCOMBAT_GUN" then
		sUnitLine = "Infantry"
	elseif sCombatClass == "UNITCOMBAT_MOUNTED" or sCombatClass == "UNITCOMBAT_ARMOR" then
		sUnitLine = "Cavalry"
	---- now we check the rest based on database order
	elseif sCombatClass == "UNITCOMBAT_RECON" then
		sUnitLine = "Recon"
	elseif sCombatClass == "UNITCOMBAT_SIEGE" then
		sUnitLine = "Siege"
	elseif sCombatClass == "UNITCOMBAT_NAVALRANGED" then
		sUnitLine = "Naval Ranged"
	elseif sCombatClass == "UNITCOMBAT_NAVALMELEE" then
		sUnitLine = "Naval Melee"
	elseif sCombatClass == "UNITCOMBAT_SUBMARINE" then
		sUnitLine = "Submarine"
	elseif sCombatClass == "UNITCOMBAT_CARRIER" then
		sUnitLine = "Carrier"
	elseif sCombatClass == "UNITCOMBAT_SETTLER" then
		sUnitLine = "Settler"
	elseif sCombatClass == "UNITCOMBAT_DIPLOMACY" then
		sUnitLine = "Diplomat"
	---- air units last, they are the rarest
	elseif sCombatClass == "UNITCOMBAT_FIGHTER" then
		sUnitLine = "Fighter"
	elseif sCombatClass == "UNITCOMBAT_BOMBER" then
		sUnitLine = "Bomber"
	elseif sCombatClass == "UNITCOMBAT_HELICOPTER" then -- unused in VP
		sUnitLine = "Helicopter"
	---- if its a combat class we did not expect, then just use the combat class type
	---- most likely will be one of the civilian types, which should not have an upgrade line
	else
		sUnitLine = sCombatClass
	end
	
	return sUnitLine
end

function Unit_DoUniqueUnitLinePromotions(iPlayer, iUnit, tPromotionsToAdd, tPromotionsToRemove)
	local pPlayer = Players[iPlayer]
	local pTeam = Teams[pPlayer:GetTeam()]
	local iCivilization = pPlayer:GetCivilizationType()
	if iCivilization == iCivilizationBarbarian or iCivilization == iCivilizationMinor then return end

	local pUnit = pPlayer:GetUnitByID(iUnit)

	-- get info about the unit
	local iCombatClass = pUnit:GetUnitCombatType()
	local sCombatClass = UnitCombat_GetTypeStringFromId(iCombatClass)
	local bMounted = pUnit:IsMounted()
	
	-- determine unit line
	local sUnitLine = Unit_GetUnitLine(sCombatClass, bMounted)

	if sUnitLine ~= nil then
		-- add unique promotions
		if tPromotionsToAdd[sUnitLine] ~= nil then
			for i, tPromotion in ipairs(tPromotionsToAdd[sUnitLine]) do
				if tPromotion[2] == nil or tPromotion[2] == "" or tPromotion[2] == -1 or pTeam:IsHasTech(tPromotion[2]) then
					pUnit:SetHasPromotion(tPromotion[1], true)
				end
			end
		end
		
		-- remove default promotions
		if tPromotionsToRemove[sUnitLine] ~= nil then
			for i, tPromotion in ipairs(tPromotionsToRemove[sUnitLine]) do
				if tPromotion[2] == nil or tPromotion[2] == "" or tPromotion[2] == -1 or pTeam:IsHasTech(tPromotion[2]) then
					pUnit:SetHasPromotion(tPromotion[1], false)
				end
			end
		end
		
		-- special cases are below
		---- Byzantium mounted units, but not armour units, has a -25% malus against cities rather than the standard -33%, and is lost on upgrade
		if iCivilization == iCivilizationByzantium and iCombatClass == iCombatClassMounted and pTeam:IsHasTech(iTechChivalry) then
			pUnit:SetHasPromotion(iPromotionSmallCityPenalty, true)
		end
	end
end
--------------------------------------------------------------------------------------------------------------------------
-- UNITCOMBATS
--------------------------------------------------------------------------------------------------------------------------
function UnitCombat_GetTypeStringFromId(iUnitCombat)
	return g_tUnitCombats[iUnitCombat]
end
--------------------------------------------------------------------------------------------------------------------------
-- UNITCLASSES
--------------------------------------------------------------------------------------------------------------------------
function UnitClass_GetTypeStringFromId(iUnitClass)
	return g_tUnitClasses[iUnitClass]
end
--=======================================================================================================================
--=======================================================================================================================